keras/models.py

```diff
@@ -211,67 +212,68 @@ def fit(self, X, y, batch_size=128, nb_epoch=100, verbose=1,
-                y_batch = y[batch_ids]
+                callbacks.on_batch_begin(batch_index)
 
-                ins = X_batch + [y_batch]
-                if show_accuracy:
-                    loss, acc = self._train_with_acc(*ins)
-                    log_values = [('loss', loss), ('acc.', acc)]
-                    av_loss += loss * len(batch_ids)
-                    av_acc += acc * len(batch_ids)
-                else:
-                    loss = self._train(*ins)
-                    log_values = [('loss', loss)]
-                    av_loss += loss * len(batch_ids)
+                try:
+                    batch_ids = index_array[batch_start:batch_end]
+                    X_batch = slice_X(X, batch_ids)
+                    y_batch = y[batch_ids]
 
-                # validation
-                if do_validation and (batch_index == len(batches) - 1):
+                    ins = X_batch + [y_batch]
                     if show_accuracy:
-                        val_loss, val_acc = self.evaluate(X_val, y_val, batch_size=batch_size, verbose=0, show_accuracy=True)
-                        log_values += [('val. loss', val_loss), ('val. acc.', val_acc)]
+                        loss, acc = self._train_with_acc(*ins)
                     else:
-                        val_loss = self.evaluate(X_val, y_val, batch_size=batch_size, verbose=0)
-                        log_values += [('val. loss', val_loss)]
-                
-                # logging
-                if verbose:
-                    progbar.update(batch_end, log_values)
-
-            history['epoch'].append(epoch)
-            history['loss'].append(av_loss/seen)
+                        loss = self._train(*ins)
+                        acc = None
+                except KeyboardInterrupt:
+                    # If training is aborted, call the callbacks anyway before terminating
+                    callbacks.on_batch_end(batch_index, [], 0., 0.)
+                    callbacks.on_epoch_end(epoch, 0., 0.)
+                    callbacks.on_train_end()
+                    raise KeyboardInterrupt # TODO: Raise a more explicit Exception (?)
+
+                callbacks.on_batch_end(batch_index, batch_ids, loss, acc)
+            
+            # validation
+            val_loss, val_acc = None, None
             if do_validation:
-                history['val_loss'].append(float(val_loss))
-            if show_accuracy:
-                history['acc'].append(av_acc/seen)
-                if do_validation:
-                    history['val_acc'].append(float(val_acc))
-        return history
+                if show_accuracy:
+                    val_loss, val_acc = self.evaluate(X_val, y_val, batch_size=batch_size, \
+                        verbose=0, show_accuracy=True)
+                else:
+                    val_loss = self.evaluate(X_val, y_val, batch_size=batch_size, verbose=0)
+
+            callbacks.on_epoch_end(epoch, val_loss, val_acc)
+
```

Comment:
Something worth noting as well is that I changed where `evaluate()` is called for validation. At the moment validation occurs at the batch level, but only for the last batch. Here validation is at the epoch level (which is equivalent to have it at the last batch), so that only `on_epoch_end()` has access to validation scores (and not `on_batch_end()`.

However it breaks the logs for validation scores since the logger update is called by `on_batch_end()` (and updating it by `on_batch_end()` and `on_epoch_end()` prints two lines somehow, I did not deeply investigate why though). So at the moment `keras.callbacks.Logger` does not display validation scores (which is the only difference with the current logging system).